### PR TITLE
create more tests and Review model and Reviews table

### PR DESCRIPTION
**Pull request by durrellchamorro**
1) While I watched the solution I found Kevin built it a little different than I did. I'd rather move on unless you think I should go back and build it like him. Here are the main differences:
I used a non model backed form, routes are a little different, instead of the create action in the Reviews controller, I made a review action in the Videos controller.

2) I couldn’t figure out how to get the flash styling, so the flash is just plain text. Any help on that would be great.

3) In the video Kevin used the syntax 
```order: “created_at DESC” ```
but that didn’t work for my version of rails. I had to use:
```-> { order(created_at: :desc) }```
In the video_spec.rb I had to write the test like this as well. Is this testing the order is descending?
  ```it { should have_many(:reviews).order(created_at: :desc) }```

4) I couldn’t figure out how to get the ```Fabricate.attributes_for``` syntax to work, so I just wrote the test like this:
```post :review, rating: review.rating, content: review.content, video: flinstones, user: current_user```
I tried:
```post :review, Fabricate.attributes_for(:review), video: flinstones ```
I put a binding.pry in the review action and all the params were filled in so I’m not sure why this didn’t work. Even when I wrote it with ```video_id: flintiness_id``` I kept getting this error in Rspec:
"Couldn't find Video without an ID”

### CODE REVIEW

File: app/controllers/videos_controller.rb [hunk 1]
```diff
@@ -7,9 +7,23 @@ def index
...
 
   def search
     @results = Video.search_by_title(params[:search_term])
   end
+
+  def review
+    @video = Video.find(params[:video])
+    review = Review.new(user_id: params[:user], video_id: params[:video], content: params[:content], rating: params[:rating])
+    if review.save
+      @reviews = @video.reviews
```
Comment (brandoncc) [inline app/controllers/videos_controller.rb]: You don't need this line. You lose the instance variable when you redirect anyway.

File: app/controllers/videos_controller.rb [hunk 2]
```diff
@@ -7,9 +7,23 @@ def index
...
+      redirect_to @video
+    else
+      @reviews = @video.reviews
+      flash[:notice] = "You must write a review"
```
Comment (brandoncc) [inline app/controllers/videos_controller.rb]: You should use `flash.now` when rendering. If you use `flash`, the message sticks around through one extra request. The next link you click on, the message will still show. Also, the reason your styling isn't working is we are using bootstrap 3 in this project. You should use "success" and "danger" for green and red alerts. You can see all of the classes in the bootstrap documentation.

File: app/views/videos/show.html.haml [hunk 3]
```diff
@@ -12,3 +12,35 @@
...
+            .row
+              .col-sm-3
+                = select_tag(:rating, options_for_select([['5 Stars', 5], ['4 Stars', 4], ["3 Stars", 3], ["2 Stars", 2], ["1 Star", 1]]), class: "form-control")
+          .form-group
+            = label_tag :content, "Write Review"
+            .row
+              .col-sm-8
+                = text_area_tag(:content, "", size: "50x6")
+          = hidden_field_tag :video, @video.id
+          = hidden_field_tag :user, current_user.id
```
Comment (brandoncc) [inline app/views/videos/show.html.haml]: Why do you need to submit this in the form? Your controller has access to the `current_user` method directly.

File: spec/controllers/application_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
+require 'spec_helper'
+
+describe ApplicationController do
```
Comment (brandoncc) [inline spec/controllers/application_controller_spec.rb]: Nice, you are the only person I have ever seen write tests for these methods. :smile:

Comment (brandoncc) [inline spec/controllers/application_controller_spec.rb]: That's a good thing, although I think the smiley looks sarcastic!

File: spec/controllers/categories_controller_spec.rb [hunk 5]
```diff
@@ -0,0 +1,13 @@
+require "spec_helper"
+
+describe CategoriesController do
+  describe "GET show" do
+    it "sets the @category variable" do
+      drama = Category.create(name: "Drama")
+      user = User.create(full_name: "bob", email_address: "bob@gmail.com", password: "password")
+      session[:user_id] = user.id
+      get :show, :id => drama.id
+      assigns(:category).should == drama
```
Comment (brandoncc) [inline spec/controllers/categories_controller_spec.rb]: Rspec recommends that you use the "expect" syntax. The "should" syntax should be considered deprecated and may even be moved to an external gem in the future.

File: spec/controllers/sessions_controller_spec.rb [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+      get :new
+      expect(response).to render_template :new
+    end
+  end
+
+  describe "POST create" do
+    context "user authenticates" do
+      it "redirects to home_path if the user authenticates" do
+        user = Fabricate(:user)
+        post :create, :email_address => user.email_address, :full_name => user.full_name, :password => user.password
```
Comment (brandoncc) [inline spec/controllers/sessions_controller_spec.rb]: You can also use the Ruby 1.9+ syntax `email_address: user.email_address` for these hashes.

File: spec/controllers/sessions_controller_spec.rb [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        expect(session[:user_id]).to eq(user.id)
+      end
+      it "sets the notice" do
+        user = Fabricate(:user)
+        post :create, :email_address => user.email_address, :full_name => user.full_name, :password => user.password
+        expect(flash[:notice]).not_to be_blank
+      end
+    end
```
Comment (brandoncc) [inline spec/controllers/sessions_controller_spec.rb]: These specs are really hard to follow. You should put a blank line between each of them to break them up. Make your code easy to read, future you will appreciate it!

File: spec/controllers/users_controller_spec.rb [hunk 8]
```diff
@@ -0,0 +1,44 @@
+require "spec_helper"
+
+describe UsersController do
+  describe "GET new" do
+    new_user = User.new
+    it "sets the @user variable" do
+      get :new
+      assigns(:user).should be_new_record
+      assigns(:user).should be_instance_of(User)
+      assigns(:user).kind_of?(User).should be_truthy
```
Comment (brandoncc) [inline spec/controllers/users_controller_spec.rb]: You can combine all three of these specs into one using the `be_a_new` matcher. Also, `be_instance_of` and `kind_of?` are the same thing. Be careful of over testing. It won't gain you any benefits and will eventually make your test suite unwieldy and slow.

File: spec/controllers/videos_controller_spec.rb [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+        review2 = Fabricate(:review, video: flinstones)
+        get :show, id: flinstones.id
+        assigns(:reviews).should =~ [review1, review2]
+      end
+
+    end
+    context "with unauthenticated user" do
+      it "redirects to login path" do
+        get :show, :id => flinstones.id
+        expect(response).to redirect_to(root_path)
```
Comment (brandoncc) [inline spec/controllers/videos_controller_spec.rb]: Your spec does not match the description. Be careful of out of sync spec descriptions.

Comment (brandoncc) [inline spec/controllers/videos_controller_spec.rb]: To be more clear, you are checking for a redirect to `root_path` but your description says `login_path`.

File: spec/fabricators/myflix_fabricator.rb [hunk 10]
```diff
@@ -0,0 +1,15 @@
+Fabricator(:user) do
```
Comment (brandoncc) [inline spec/fabricators/myflix_fabricator.rb]: I would recommend breaking these up into their own files like the solutions. It makes them much easier to find and work with in the long run.

**Comment by brandoncc** [comment]
1. Since a review is actually a resource with a model, it stands to reason that it should be created using a model backed form. Resources are usually dealt with using Restful routing conventions as the solutions do. I would recommend getting used to doing it that way because if you ever work on an app which utilizes a front end framework your backend will be required to use Restful routing conventions.

2. As I mentioned in one of the comments, this is because the class names changed in bootstrap 3.

3. The course was based on Rails 3.2 originally. Rails 4 introduced the new syntax which you had to use.

4. The key for the video is `id` not `video`. Whatever param key you are using in your controller is the one you have to assign in your spec.

File: app/controllers/videos_controller.rb (hunk 1 above)
Comment (durrellchamorro) [inline app/controllers/videos_controller.rb]: Since the view needs ```@reviews``` to load, why does not setting ```@reviews``` in the controller work?

Comment (brandoncc) [inline app/controllers/videos_controller.rb]: You are not rendering a view, you are redirecting. The `videos#show` action does the rendering. When you redirect, it is a whole new request/response cycle. Since HTTP is stateless, you lose the ivars.
